server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -0,0 +1,128 @@
+import com.cloud.vm.dao.VMInstanceDao;
+import org.apache.cloudstack.api.command.admin.diagnostics.ExecuteDiagnosticsCmd;
+import org.apache.cloudstack.engine.orchestration.service.NetworkOrchestrationService;
+import org.apache.log4j.Logger;
+
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class DiagnosticsServiceImpl extends ManagerBase implements PluggableService, DiagnosticsService {
+    private static final Logger LOGGER = Logger.getLogger(DiagnosticsServiceImpl.class);
+
+    @Inject
+    private RouterControlHelper routerControlHelper;
+    @Inject
+    private AgentManager agentManager;
+    @Inject
+    private VMInstanceDao instanceDao;
+    @Inject
+    private VirtualMachineManager vmManager;
+    @Inject
+    private NetworkOrchestrationService networkManager;
+
+    @Override
+    public Map<String, String> runDiagnosticsCommand(final ExecuteDiagnosticsCmd cmd) throws AgentUnavailableException, InvalidParameterValueException {
+        final Long vmId = cmd.getId();
+        final String cmdType = cmd.getType().getValue();
+        final String cmdAddress = cmd.getAddress();
+        final String optionalArguments = cmd.getOptionalArguments();
+        final VMInstanceVO vmInstance = instanceDao.findByIdTypes(vmId, VirtualMachine.Type.ConsoleProxy, VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+
+        if (vmInstance == null) {
+            throw new InvalidParameterValueException("Unable to find a system vm with id " + vmId);
+        }
+        final Long hostId = vmInstance.getHostId();
+
+        if (hostId == null) {
+            throw new CloudRuntimeException("Unable to find host for virtual machine instance: " + vmInstance.getInstanceName());
+        }
+
+        final String shellCmd = prepareShellCmd(String.format("%s %s %s", cmdType, cmdAddress, optionalArguments));
+
+        final DiagnosticsCommand command = new DiagnosticsCommand(shellCmd, vmManager.getExecuteInSequence(vmInstance.getHypervisorType()));
+        final Map<String, String> accessDetails = networkManager.getSystemVMAccessDetails(vmInstance);
+        command.setAccessDetail(accessDetails);
+
+        Answer answer =  agentManager.easySend(hostId, command);
+        final Map<String, String> detailsMap = ((DiagnosticsAnswer) answer).getExecutionDetails();
+
+        if (!detailsMap.isEmpty()) {
+            return detailsMap;
+        } else {
+            throw new CloudRuntimeException("Error occurred when executing diagnostics command on remote target: " + answer.getDetails());
+        }
+    }
+
+    protected String prepareShellCmd(CharSequence cmdArguments) {
+        List<String> tokens = new ArrayList<>();
+        boolean escaping = false;
+        boolean quoting = false;
+        char quoteChar = ' ';
+        StringBuilder current = new StringBuilder();
+
+        for (int i = 0; i < cmdArguments.length(); i++) {
```

Comment:
See if a lex-parsing library could be used instead. It's not a good idea to write C-styled code in java for manual parsing, the code is hard to understand. If you choose to keep it as such, add javadoc to explain the algorithm.